spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/Spark3ZOrderUDF.java

```diff
@@ -0,0 +1,241 @@
+import org.apache.spark.sql.functions;
+import org.apache.spark.sql.types.BinaryType;
+import org.apache.spark.sql.types.BooleanType;
+import org.apache.spark.sql.types.ByteType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.DateType;
+import org.apache.spark.sql.types.DoubleType;
+import org.apache.spark.sql.types.FloatType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.ShortType;
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.TimestampType;
+import scala.collection.Seq;
+
+class SparkZOrder implements Serializable {
+    private final int STRING_KEY_LENGTH = 128;
+
+    private final byte[] TINY_EMPTY = new byte[Byte.BYTES];
+    private final byte[] SHORT_EMPTY = new byte[Short.BYTES];
+    private final byte[] INT_EMPTY = new byte[Integer.BYTES];
+    private final byte[] LONG_EMPTY = new byte[Long.BYTES];
+    private final byte[] FLOAT_EMPTY = new byte[Float.BYTES];
+    private final byte[] DOUBLE_EMPTY = new byte[Double.BYTES];
+
+    transient private ThreadLocal<ByteBuffer> outputBuffer;
+    transient private ThreadLocal<byte[][]> inputHolder;
+    transient private ThreadLocal<ByteBuffer>[] inputBuffers;
+    transient private ThreadLocal<CharsetEncoder> encoder;
+
+    private final int numCols;
+
+    private int inputCol = 0;
+    private int totalBytes = 0;
+
+    SparkZOrder(int numCols) {
+        this.numCols = numCols;
+    }
+
+    private void readObject(java.io.ObjectInputStream in)
+            throws IOException, ClassNotFoundException {
+        in.defaultReadObject();
+        inputBuffers = new ThreadLocal[numCols];
+        inputHolder = ThreadLocal.withInitial(() -> new byte[numCols][]);
+        encoder = ThreadLocal.withInitial(() -> StandardCharsets.UTF_8.newEncoder());
+    }
+
+
+    private ByteBuffer outputBuffer(int size) {
+        if (outputBuffer == null) {
+            // May over allocate on concurrent calls
+            outputBuffer = ThreadLocal.withInitial(() -> ByteBuffer.allocate(size));
+        }
+        return outputBuffer.get();
+    }
+
+    private ByteBuffer inputBuffer(int position, int size){
+        if (inputBuffers[position] == null) {
+            // May over allocate on concurrent calls
+            inputBuffers[position] = ThreadLocal.withInitial(() -> ByteBuffer.allocate(size));
+        }
+        return inputBuffers[position].get();
+    }
+
+    byte[] interleaveBits(Seq<byte[]> scalaBinary) {
+        byte[][] columnsBinary = scala.collection.JavaConverters.seqAsJavaList(scalaBinary)
+                .toArray(inputHolder.get());
+        return ZOrderByteUtils.interleaveBits(columnsBinary, totalBytes, outputBuffer(totalBytes));
+    }
+
+    private UserDefinedFunction tinyToOrderedBytesUDF() {
+        int position = inputCol;
```

Comment:
Nit: could this repetitive boilerplate be extracted into a method, ie:

```
private UserDefinedFunction typeToOrderedBytesUDF<T> (byte[] empty, int bytes, String name)
```

Reply:
Do you have any ideas how this would look? I'm not sure we can make it much better ... but maybe I lack vision

Reply:
Sorry, you are right, you can ignore this comment.